### PR TITLE
DAOS-3114 vos: Remove VOS_OF_USE_TIMESTAMPS flag, replace with dth

### PR DESCRIPTION
**Pull request by jolivier23**
This limits any timestamp checks and updates to cases where a dtx
handle exists.  It updates the mvcc and pm tests to use transactions
where needed to get timestamp updates and checks.  It also adds the
check for same transaction updates by comparing the transaction uuid.
Adds an internal API to get unique uuid from the same thread for
internal purposes.   Adds ex API for vos_obj_update/fetch to pass
dth

Updates vos_ts.h to Doxygen style

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### CODE REVIEW

File: src/include/daos_srv/vos_types.h [hunk 1]
```diff
@@ -214,15 +214,14 @@ enum {
 	/** Conditional Op: Fetch akey if it exists, fail otherwise */
 	VOS_OF_COND_AKEY_FETCH	= DAOS_COND_AKEY_FETCH,
 	/** Indicates the operation should check mvcc timestamps */
```
Comment (liw) [inline src/include/daos_srv/vos_types.h:216]: Forgot to remove this comment along with VOS_OF_USE_TIMESTAMPS?

File: src/vos/vos_ts.h [hunk 2]
```diff
@@ -419,82 +425,91 @@ vos_ts_evict(uint32_t *idx, uint32_t type)
...
+ *
+ * \param[in]	entry		The timestamp entry
+ * \param[in]	read_time	The new read timestamp
+ * \param[in]	tx_id		The uuid of the new read
+ */
+static inline void
+vos_ts_rl_update(struct vos_ts_entry *entry, daos_epoch_t read_time,
+		 const uuid_t tx_id)
+{
+	if (entry == NULL || read_time <= entry->te_ts_rl)
```
Comment (liw) [inline src/vos/vos_ts.h:485]: If read_time = te_ts_rl but tx_id != te_tx_rl, then we should assign tx_id to te_tx_rl, no?

File: src/vos/vos_ts.h [hunk 3]
```diff
@@ -419,82 +425,91 @@ vos_ts_evict(uint32_t *idx, uint32_t type)
...
+ *
+ * \param[in]	entry		The timestamp entry
+ * \param[in]	read_time	The new read timestamp
+ * \param[in]	tx_id		The uuid of the new read
+ */
+static inline void
+vos_ts_rh_update(struct vos_ts_entry *entry, daos_epoch_t read_time,
+		 const uuid_t tx_id)
+{
+	if (entry == NULL || read_time <= entry->te_ts_rh)
```
Comment (liw) [inline src/vos/vos_ts.h:502]: Ditto for "=" in "<=".

File: src/include/daos_srv/vos.h [hunk 4]
```diff
@@ -425,6 +425,35 @@ vos_obj_fetch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
 	      uint64_t flags, daos_key_t *dkey, unsigned int iod_nr,
 	      daos_iod_t *iods, d_sg_list_t *sgls);
 
+/**
+ * Fetch values for the given keys and their indices.
+ * If output buffer is not provided in \a sgl, then this function returns
+ * the directly accessible addresses of record data, upper layer can directly
+ * read from these addresses (rdma mode).
+ *
+ * TODO: add more detail descriptions for punched or missing records.
```
Comment (ryon-jensen) [inline src/include/daos_srv/vos.h:434]: Why not do this TODO now?

File: src/include/daos_srv/vos.h [hunk 5]
```diff
@@ -425,6 +425,35 @@ vos_obj_fetch(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
...
+ *
+ * \param coh	[IN]	Container open handle
```
Comment (ryon-jensen) [inline src/include/daos_srv/vos.h:436]: Effort was made to update vos_ts.h to move [IN] to before param name. Is there a reason to have them before vs after? Also uper vs lowercase? No big deal, just wondering about consistency.

File: src/vos/vos_ts.h (hunk 2 above)
Comment (jolivier23) [reply src/vos/vos_ts.h:485]: yes, perhaps we should.   Though it's a little unclear to me.   If two transactions have same timestamp and both try to modify the same key, one will be asked to restart.

File: src/include/daos_srv/vos.h (hunk 4 above)
Comment (jolivier23) [reply src/include/daos_srv/vos.h:434]: I didn't add this TODO.   I probably need to clean up the vos header at some point but don't want to do it here.

File: src/include/daos_srv/vos.h (hunk 5 above)
Comment (jolivier23) [reply src/include/daos_srv/vos.h:436]: \param[in] is correct for Doxygen.   So I'm opportunistically updating files.

File: src/vos/vos_ts.h (hunk 2 above)
Comment (liw) [reply src/vos/vos_ts.h]: Yeah, this requires a little deduction from the rules in src/vos/README.md. Consider:

    TX1 reads  value x at 10: x.low = {10, TX1}
    TX2 reads  value x at 10: x.low = {10, TX2} <-- required to reject next line
    TX1 writes value x at 10: reject due to TX ID differences

**Comment by jolivier23** [comment]
DAOS-4537
DAOS-5295

**Pull request merged**